python/nano/src/bigdl/nano/tf/optimizers/sparse_adam.py

```diff
@@ -0,0 +1,134 @@
+# Copyright 2016 The BigDL Authors.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ==============================================================================
+"""SparseAdam optimizer implementation."""
+
+import tensorflow.compat.v2 as tf
+from keras import backend_config
+from keras.optimizer_v2 import optimizer_v2
+from tensorflow.python.util.tf_export import keras_export
+import tensorflow
+
+
+class SparseAdam(tensorflow.keras.optimizers.Adam):
+    """
+    Variant of the Adam optimizer that handles sparse updates more
+    efficiently.
+
+    The original Adam algorithm maintains two moving-average accumulators for
+    each trainable variable; the accumulators are updated at every step.
+    In this variant, only moments that show up in the gradient get updated,
+    and only those portions of the gradient get applied to the parameters.
+    Compared with the original Adam optimizer, it can provide large improvements in
+    model training throughput for some applications.
+
+    Note, :pram amsgrad is currently not supported and the argument can only be set to
+    False.
+    """
+    _HAS_AGGREGATE_GRAD = True
+
+    def __init__(self,
+                 learning_rate=0.001,
+                 beta_1=0.9,
+                 beta_2=0.999,
+                 epsilon=1e-7,
+                 amsgrad=False,
+                 name='SparseAdam',
+                 **kwargs):
+        '''
+        This is a slightly modified version of tf.keras.optimizers.Adam,
+        which only apply resource_apply_sparse to the output of the embedding

```

Comment:
"which only update moving-average accumulators for sparse variable indices that appear in the current batch"